server/src/main/java/org/opensearch/index/mapper/NumberFieldMapper.java

```diff
@@ -201,9 +205,14 @@ public Float parse(XContentParser parser, boolean coerce) throws IOException {
             }
 
             @Override
-            public Query termQuery(String field, Object value) {
+            public Query termQuery(String field, Object value, Boolean hasDocValues) {
                 float v = parse(value, false);
-                return HalfFloatPoint.newExactQuery(field, v);
+                Query query = HalfFloatPoint.newExactQuery(field, v);
+                if (hasDocValues) {
+                    Query dvQuery = SortedNumericDocValuesField.newSlowExactQuery(field, HalfFloatPoint.halfFloatToSortableShort(v));
+                    query = new IndexOrDocValuesQuery(query, dvQuery);
+                }
+                return query;
```

Comment:
Similar to what you did for keyword queries in https://github.com/opensearch-project/OpenSearch/pull/10425, can you please handle all four possible combinations of indexed + docvalues:

1. **indexed, no doc values**, use the `<Whatever>Point.new<Thing>Query()`.
2. **doc values, not indexed**, use the `SortedNumericDocValuesField.newSlow<Thing>Query()`
3. **both**: Do `IndexOrDocValuesQuery` over both the above queries.
4. **neither**: Throw an exception.

Reply:
In particular, you made the change to call `failIfNotIndexedAndNoDocValues()`, but right now you're always producing the index query, so you're not covering case 2.

Reply:
You're right. I've addressed this.